### ISSUE TITLE
Consider not recommending use of virtual environments for daily use

### ISSUE THREAD

**Issue by chrisjbillington** [labscript-suite/labscript-suite#45]
tl;dr: Thinking about users installing the labscript suite into virtual environments and non-base conda environments presents a few ways things can go wrong that push away from the 'working out of the box' experience. Whilst these are mostly fixable and make sense if you are an experienced pythonista, I don't believe these impediments are worth it for the benefits of using isolated environments. I don't think lab computers are the multi-purpose machines that use of venvs implies, and that their main purpose in running Python is to run the labscript suite, and that the labscript suite sharing an env with other packages for other purposes should be fine.

----

Virtual environments and conda envs are a great way to keep different Python contexts separate. This is great for keeping e.g. development or experimental work isolated from a daily-use Python installation. However, I'm leaning toward the position that base environments should be recommended for day-to-day lab use.

#### Integration with other tools is lacking/fiddly in non-base envs

Firstly, there is integration with text editors and other tools. One sees this message when installing miniconda/anaconda (this box is by default checked):

![image](https://user-images.githubusercontent.com/1044087/83948986-2ea7e980-a7ef-11ea-8bc1-561536ef525d.png)

And a similar message when installing Python from Python.org. Without the Python installation being in the registry, one must manually configure tools such as text editors to see it. If one installs only one Python distribution (Microsoft Store, Python.org or Anaconda/Miniconda), then installing with the defaults means that tools that look for a Python installation will 'just work'. Conda envs are only registered in conda-specific files on disk and virtual environments are not registered anywhere - tools will not know about these Python environments via the registry mechanism.

I see a lot of users using Spyder, which I think is a fine choice for many. However, getting Spyder to work with non-base environments, whilst simple once you understand it, is [several extra steps with potential for confusion](https://github.com/spyder-ide/spyder/wiki/Working-with-packages-and-environments-in-Spyder), with one possible option ending up with multiple Spyder launchers in your start menu, and another involving installing ipython kernels into envs, but running spyder from the base env still.

Jupyter notebooks are commonly used as well. Although I haven't looked into it specifically, it [looks like there is an extra step or so](https://stackoverflow.com/a/42585283/437537) to getting them working with a non-base env as well. Perhaps this step is not necessary if you always launch jupyter from a terminal, and perhaps the solution is really not that bad once you've set it up. But if you look at the rest of that stackoverflow discussion, it's clear that there are plenty of opportunities for confusion and time-wasting that simply don't exist if one uses the base environment by default. 

#### Removing a broken env and starting over is also possible in a base env

Whilst removing an env and starting over is trivially possible for non-base envs, it's not that bad for base envs as well. 

For Microsoft Store Python, all packages/scripts installed without a venv are installed to a subdirectory of `%AppData%\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0` or similar. Starting fresh comprises simply deleting this directory.

For conda, one can simply uninstall and reinstall Anaconda/Miniconda (the latter taking only minutes instead of tens of minutes for the former). [One can even keep non-base envs if they want](https://www.reddit.com/r/Python/comments/636yb3/reinstall_conda_miniconda_but_keep_envs/), though I'm imagining the simplest case where people are not using envs at all. This might sound like a pain in the neck, and it is if you're doing it every day.

For Python.org Python, unfortunately pip installs everything to the "system" site-packages directory. So to start fresh one must either uninstall and reinstall Python, or one must have been sure to use `pip install --user` consistently, in which case one can just delete  `%AppData%\Roaming\Python\Python38` similarly to store-bought Python.

This uninstalling and reinstalling might sound arduous, and it is, but mostly in the sense of taking more time. However, it is also conceptually straightforward. Whereas getting tools working with non-base Python environments I think will be a bigger time sink as it requires more research and trial and error to get right.

#### The labscript suite is a well-behaved set of packages without special requirements

The labscript suite should be able to be installed into practically any Python environment. Our dependency tree is not deep, we do not depend on any packages that themselves require for example older versions of other packages creating conflicts. We do not require conda-forge and only vendor packages not available in standard conda repos. We ensure to work with the latest versions of all our dependencies. Therefore, it is less necessary to require a special environment for the labscript suite.

I don't think labs will adopt a workflow where they put packages that need to interact with labscript in the labscript env and other packages in another env. If they follow our current recommendations for creating an environment, I suspect many users will simply always activate the labscript env and use it for everything, defeating the purpose of having it isolated.

#### Conda breakage is often not fixed by re-creating environments anyway

Much conda breakage does not have much to do with specific environments and is instead `conda` itself from the base environment being broken. Whilst this might be less likely to occur if the base environment is never updated, updating conda itself is still necessary to use new packaging features as they come out, as well as to get bugfixes. It's usually recommended to update conda in the base environment before doing anything new.

So when ones conda setup is broken, it's often the case that it requires reinstalling anaconda anyway. Which, despite the "ugh" factor of this being necessary, is not actually as arduous a step as it may seem compared to recreating an environment.

#### in summary

Virtual envs and conda envs have real benefits, but unfortunately, I don't think they are worth the added complexity for day-to-day use. We want to be compatible with a wide range of tools that users might use with Python, and recommending the use of base environments gets us decently closer to the ideal of things "just working".

I think we should recommend venvs and conda envs for development installs *if* there also exists a regular install. Even for the case of a development install, if the development install is being used day-to-day I think it is better if it lives in the base environment. Venvs and conda envs should be used to *keep other things away from the day-to-day labscript suite installation* (if necessary, which as argued above is seldom), not to keep the labscript suite away from everything else.

Our documentation should continue to mention the use of venvs and conda envs as a possibility for when you have multiple things you are using Python for that you want to keep separate, but it should be absent from the quick-start.

If we like, we can have a section on "What if I break my Python installation" that mentions the uninstall/reinstall options above. I would contend that actually for the case of conda and Python.org, users unfamiliar with the details will default to uninstall/reinstall naturally. Whereas actually recreating an env is the more counter intuitive thing, and since Microsoft Store Python persists user-installed data across installs, that will not occur with a simple uninstall/reinstall. Though if there is a "delete app data" option for Microsoft Store apps, that would probably do the trick.

**Comment by chrisjbillington** [comment]
For what it's worth VS Code seems to handle conda envs quite well (in that it can list them automatically and activate them to run Python code - I haven't actually tried it out but it looks very nice). Though @philipstarkey mentioned on Zulip having difficulty with a VS code extension that was not playing nice with a conda env until saying the appropriate magic words. Perhaps the extension would have had just as much trouble with the base conda env as with a non-base one, but this is kind of issue I think is quite widespread at the moment in text editors and IDEs.

**Comment by chrisjbillington** [comment]
Some of these arguments also would seem to point in the direction of recommending regular Store/.org Python over Conda since conda envs are more difficult to activate programmatically  (regular Python without a venv doesn't require activation of any kind, and neither do venvs so long as you can point to the interpreter executable). However, the most common tools like Spyder and Jupyter do work well with conda, and I would actually expect them to work better in conda than elsewhere (unjustified, but just my intuition since they're firmly in that ecosystem). Other tools don't know how to activate a conda env, but on balance I think conda still wins out for now.

I'm very much guided by trying to match the spyder/jupyter in a base conda env setup that the large majority of people using Python in physics around me are using. These tools are clearly filling a niche and even though I don't use them I would like our default recommendations to fit within that ecosystem with as little friction as possible.

**Comment by philipstarkey** [comment]
So I've been thinking about this a bit...here are the main thoughts I have:

- Much of this seems to be predicated on the assumption that people are not already using Python for something else. Installing the labscript suite (and thus newer versions of dependencies) in a base environment could break existing workflows unrelated to the labscript suite. For example, a little while ago an upgrade of matplotlib (probably v1 to v2?) completely stuffed the formatting of graphs produced by a bunch of scripts I was using. I would rather our installation instructions were not responsible for breaking other Python software they might already be using.

- Are people really using other tools with the labscript suite? Even if they do use an IDE (Spyder/Sublime/VS Code/etc.), it's just for code editing right? Experiment logic files and analysis scripts aren't typically executed from the IDE, they're executed by labscript suite components. Maybe on a personal PC if you write your analysis script to also run outside of lyse? but that brings it's own complexities anyway.

- Even if people are using different tools, there is likely a wide variety in the combination of tools in use. For example, just at Monash the spinor lab uses Sublime + Conda envs (IIRC) and the KRb labs use conda base env + notepad++. I thus feel like suggesting the default installation procedure not virtual envs, because some tools don't play nice, isn't a particularly convincing argument because it will always be true for some set of tools people are using, and we don't have any data on what people are currently using outside of our immediate experiences.

(I feel like I had another point I wanted to make, but I got interrupted halfway through writing this and I'm blanking now).

I personally think our installation instructions should aim for doing as minimal "damage" as possible to a users existing setup. We want people to experiment with our software by installing it and playing around without their being any consequences for their existing setup. Concern about breaking what currently works in the lab is already an impediment to getting people to try out our software. This means advocating for virtual environments in the installation instructions.

P.S. If people are using Jupyter/Spyder they should really consider trying VS Code. As you say, it has very good interaction with conda environments (except for the occasional 3rd party plugin). It has the ability to edit and run .ipynb files in a Jupyter style interface and other features similar to Spyder.

**Comment by rpanderson** [comment]
Interesting discussion. Aside from editor/IDE/notebook integration (discussed further below), I'm missing how the other points fit into the category of virtual environments compromising things 'working out of the box'. Is the idea of "(using) of base environments (to get) us decently closer to the ideal of things 'just working'" actually more about convenience?

>  I don't think lab computers are the multi-purpose machines that use of venvs implies, and that their main purpose in running Python is to run the labscript suite, and that the labscript suite sharing an env with other packages for other purposes should be fine.

This presumes that the labscript suite will be installed on (or at least our installation docs should target these exclusively, predominantly), yet one of the main features of the labscript suite is the ability to install different components on different hosts, which need not be laboratory hosts.

Notwithstanding this, it doesn't take more than two purposes (or packages) to warrant the use of virtual environments. 

> *The labscript suite is a well-behaved set of packages without special requirements*

What if a lab machine is used for e.g. image analysis, simulation, or machine learning, asynchronous to the labscript suite? Just because our dependency tree is not deep and we don't depend on older versions of other packages, isn't to say that another package might require an older version, preventing us from requiring a minimum one.

> I think we should recommend venvs and conda envs for development installs if there also exists a regular install. Even for the case of a development install, if the development install is being used day-to-day I think it is better if it lives in the base environment. Venvs and conda envs should be used to _keep other things away from the day-to-day labscript suite installation_ (if necessary, which as argued above is seldom), not to keep the labscript suite away from everything else.

Virtual environments are about keeping things away from each other, not centred around the labscript suite. The above presumes predominantly (day-to-day?) labscript Python usage. New users might not immediately be day-to-day labscript users, initially.

> *Integration with other tools is lacking/fiddly in non-base envs*

Per @philipstarkey's points, is there a use case for requiring an editor to be an IDE for labscript suite components? Or for using Jupyter notebooks? I'm actually in favour of both, but can only think of these being applied to lyse analysis.

> For what it's worth VS Code seems to handle conda envs quite well (in that it can list them automatically and activate them to run Python code - I haven't actually tried it out but it looks very nice). Though @philipstarkey mentioned on Zulip having difficulty with a VS code extension that was not playing nice with a conda env until saying the appropriate magic words. Perhaps the extension would have had just as much trouble with the base conda env as with a non-base one, but this is kind of issue I think is quite widespread at the moment in text editors and IDEs.

My experience of VS Code has been it solving this pretty well out of the box:
![image](https://user-images.githubusercontent.com/5637107/84091022-a61b8a00-aa36-11ea-9589-598a30352756.png)

Anaconda environments, Microsoft Store Python, and local .venv are all detected. This is using the [Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.python) and [Anaconda Extension Pack](https://marketplace.visualstudio.com/items?itemName=ms-python.anaconda-extension-pack).
